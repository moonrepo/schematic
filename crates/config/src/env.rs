use crate::error::ConfigError;
use std::str::FromStr;

/// Parse a string into a boolean. Will parse `1`, `true`, `yes`, `on`,
/// and `enabled` as true, and everything else as false.
pub fn parse_bool(var: String) -> Result<bool, ConfigError> {
    match var.to_lowercase().as_str() {
        "1" | "true" | "yes" | "on" | "enabled" | "enable" => Ok(true),
        _ => Ok(false),
    }
}

fn split<T: FromStr>(var: String, delim: char) -> Result<Vec<T>, ConfigError> {
    let mut list = vec![];

    for s in var.split(delim) {
        let value: T = s.trim().parse().map_err(|_| {
            ConfigError::Message(format!("Failed to parse \"{s}\" into the correct type."))
        })?;

        list.push(value);
    }

    Ok(list)
}

/// Split a variable on each comma (`,`) and parse into a list of values.
pub fn split_comma<T: FromStr>(var: String) -> Result<Vec<T>, ConfigError> {
    split(var, ',')
}

/// Split a variable on each colon (`:`) and parse into a list of values.
pub fn split_colon<T: FromStr>(var: String) -> Result<Vec<T>, ConfigError> {
    split(var, ':')
}

/// Split a variable on each semicolon (`;`) and parse into a list of values.
pub fn split_semicolon<T: FromStr>(var: String) -> Result<Vec<T>, ConfigError> {
    split(var, ';')
}

/// Split a variable on each space (` `) and parse into a list of values.
pub fn split_space<T: FromStr>(var: String) -> Result<Vec<T>, ConfigError> {
    split(var, ' ')
}
