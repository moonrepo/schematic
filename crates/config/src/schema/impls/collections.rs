// use crate::config::ConfigSchema;
// use crate::schema::types::*;

// impl<T: ConfigSchema> ConfigSchema for Vec<T> {
//     fn generate_schema() -> Schema {
//         let inner = T::generate_schema();

//         if let Schema::Type {
//             ref mut nullable, ..
//         } = schema
//         {
//             *nullable = true;
//         }

//         schema
//     }
// }
