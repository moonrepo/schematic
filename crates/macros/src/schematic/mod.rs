use crate::common::Macro;
use crate::utils::extract_comment;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

pub struct SchematicMacro<'l>(pub Macro<'l>);

impl<'l> ToTokens for SchematicMacro<'l> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let cfg = &self.0;
        let name = cfg.name;
        let schema = cfg.type_of.generate_schema(
            name,
            extract_comment(&cfg.attrs),
            cfg.get_casing_format(),
            cfg.get_tagged_format(),
        );

        tokens.extend(quote! {
            #[automatically_derived]
            impl schematic::Schematic for #name {
                fn generate_schema() -> schematic::SchemaType {
                    use schematic::schema::*;
                    #schema
                }
            }
        });
    }
}
