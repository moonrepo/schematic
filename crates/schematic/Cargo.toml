[package]
name = "schematic"
version = "0.15.1"
edition = "2021"
license = "MIT"
description = "A layered serde configuration and schema library."
homepage = "https://moonrepo.github.io/schematic"
repository = "https://github.com/moonrepo/schematic"
readme = "../../README.md"

[package.metadata.release]
pre-release-replacements = [
	{ file = "../../CHANGELOG.md", search = "Unreleased", replace = "{{version}}" },
]

[package.metadata.docs.rs]
all-features = true

[dependencies]
schematic_macros = { version = "0.15.0", path = "../macros" }
schematic_types = { version = "0.7.0", path = "../types" }
miette = { workspace = true }
thiserror = "1.0.59"
tracing = "0.1.40"

# config
garde = { version = "0.18.0", default-features = false, optional = true, features = [
	"regex",
] }
serde = { workspace = true }
serde_path_to_error = { version = "0.1.16", optional = true }
starbase_styles = { version = "0.4.0", optional = true }

# schema
indexmap = { workspace = true, optional = true, features = ["serde"] }

# json
serde_json = { workspace = true, optional = true, features = [
	"preserve_order",
] }

# json schema
markdown = { version = "1.0.0-alpha.17", optional = true }
schemars = { version = "0.8.17", optional = true, default-features = false }

# toml
toml = { workspace = true, optional = true }

# yaml
serde_yaml = { workspace = true, optional = true }

# url
reqwest = { workspace = true, optional = true, features = ["blocking"] }

[features]
default = ["config", "url"]
config = [
	"dep:garde",
	"dep:serde_path_to_error",
	"dep:starbase_styles",
	"schematic_macros/config",
]
json = ["dep:serde_json"]
schema = ["dep:indexmap", "schematic_macros/schema"]
toml = ["dep:toml"]
url = ["dep:reqwest"]
yaml = ["dep:serde_yaml"]

renderer_json_schema = ["dep:markdown", "dep:schemars", "json", "schema"]
renderer_template = []
renderer_typescript = ["schema"]

type_chrono = ["schematic_types/chrono"]
type_indexmap = ["schematic_types/indexmap"]
type_regex = ["schematic_types/regex"]
type_relative_path = ["schematic_types/relative_path"]
type_rust_decimal = ["schematic_types/rust_decimal"]
type_semver = ["schematic_types/semver"]
type_serde_json = ["schematic_types/serde_json"]
type_serde_toml = ["schematic_types/serde_toml"]
type_serde_yaml = ["schematic_types/serde_yaml"]
type_url = ["schematic_types/url"]

valid_email = ["garde/email"]
valid_url = ["garde/url"]

[dev-dependencies]
schematic = { path = ".", features = [
	"config",
	"json",
	"schema",
	"toml",
	"renderer_json_schema",
	"renderer_template",
	"renderer_typescript",
	"type_chrono",
	"type_indexmap",
	"type_regex",
	"type_relative_path",
	"type_rust_decimal",
	"type_semver",
	"type_serde_json",
	"type_serde_toml",
	"type_serde_yaml",
	"type_url",
	"url",
	"valid_email",
	"valid_url",
	"yaml",
] }
reqwest = { workspace = true, features = [
	"blocking",
	"rustls-tls-native-roots",
] }
serial_test = "3.1.1"
starbase_sandbox = "0.6.0"

# Types
chrono = { workspace = true }
indexmap = { workspace = true }
regex = { workspace = true }
relative-path = { workspace = true, features = ["serde"] }
rust_decimal = { workspace = true }
semver = { workspace = true, features = ["serde"] }
url = { workspace = true, features = ["serde"] }
