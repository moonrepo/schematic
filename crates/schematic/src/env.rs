use crate::{internal, ParseEnvResult};
use std::str::FromStr;

/// Ignore the environment variable if it's empty and fallback to the previous or default value.
pub fn ignore_empty<T: FromStr>(var: String) -> ParseEnvResult<T> {
    let var = var.trim();

    if var.is_empty() {
        return Ok(None);
    }

    internal::parse_value(var).map(|v| Some(v))
}

/// Parse a string into a boolean. Will parse `1`, `true`, `yes`, `on`,
/// and `enabled` as true, and everything else as false.
pub fn parse_bool(var: String) -> ParseEnvResult<bool> {
    Ok(match var.to_lowercase().as_str() {
        "1" | "true" | "yes" | "on" | "enabled" | "enable" => Some(true),
        _ => Some(false),
    })
}

fn split<T: FromStr>(var: String, delim: char) -> ParseEnvResult<Vec<T>> {
    let mut list = vec![];

    for s in var.split(delim) {
        let value = s.trim();

        if !value.is_empty() {
            list.push(internal::parse_value(value)?);
        }
    }

    Ok(Some(list))
}

/// Split a variable on each comma (`,`) and parse into a list of values.
pub fn split_comma<T: FromStr>(var: String) -> ParseEnvResult<Vec<T>> {
    split(var, ',')
}

/// Split a variable on each colon (`:`) and parse into a list of values.
pub fn split_colon<T: FromStr>(var: String) -> ParseEnvResult<Vec<T>> {
    split(var, ':')
}

/// Split a variable on each semicolon (`;`) and parse into a list of values.
pub fn split_semicolon<T: FromStr>(var: String) -> ParseEnvResult<Vec<T>> {
    split(var, ';')
}

/// Split a variable on each space (` `) and parse into a list of values.
pub fn split_space<T: FromStr>(var: String) -> ParseEnvResult<Vec<T>> {
    split(var, ' ')
}
