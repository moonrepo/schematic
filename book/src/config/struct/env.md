# Environment variables

> Not supported for enums.

Settings can also inherit values from environment variables via the `#[setting(env)]` attribute
field. When using this, variables take the _highest_ precedence, and are merged as the last layer.

```rust
#[derive(Config)]
struct AppConfig {
	#[setting(default = 3000, env = "PORT")]
	pub port: usize,
}
```

## Container prefixes

If you'd prefer to not define `env` for _every_ setting, you can instead define a prefix on the
containing struct using the `#[setting(env_prefix)]` attribute field. This will define an
environment variable for _all_ direct fields in the struct, in the format of "env prefix + field
name" in UPPER_SNAKE_CASE.

For example, the environment variable below for `port` is now `APP_PORT`.

```rust
#[derive(Config)]
#[config(env_prefix = "APP_")]
struct AppConfig {
	#[setting(default = 3000)]
	pub port: usize,
}
```

### Nested prefixes

Since `env_prefix` only applies to direct fields and not for nested/children structs, you'll need to
define `env_prefix` for each struct, and manually set the prefixes. Schematic _does not concatenate_
the prefixes between parent and child.

```rust
#[derive(Config)]
#[config(env_prefix = "APP_SERVER_")]
struct AppServerConfig {
	// ...
}

#[derive(Config)]
#[config(env_prefix = "APP_")]
struct AppConfig {
	#[setting(nested)]
	pub server: AppServerConfig,
}
```

## Parsing values

We also support parsing environment variables into the required type. For example, the variable may
be a comma separated list of values, or a JSON string.

The `#[setting(parse_env)]` attribute field can be used, which requires a path to a function to
handle the parsing, and receives the variable value as a single argument.

```rust
#[derive(Config)]
struct AppConfig {
	#[setting(env = "ALLOWED_HOSTS", parse_env = schematic::env::split_comma)]
	pub allowed_hosts: Vec<String>,
}
```

> We provide a handful of built-in parsing functions in the
> [`env` module](https://docs.rs/schematic/latest/schematic/env/index.html).

## Parse function

You can also define your own function for parsing values out of environment variables.

When defining a custom parse function, the variable value is passed as the 1st argument, and a
`Result<Option>` must be returned. A `None` value can be returned, which will fallback to the
previous or default value.

```rust
use schematic::ConfigError;

pub fn custom_parse(var: String) -> Result<Some<ReturnValue>, ConfigError> {
	do_parse()
		.map(|v| Some(v))
		.map_err(|e| ConfigError::Message(e.to_string()))
}
```

If parsing fails, you must return a
[`ConfigError`](https://docs.rs/schematic/latest/schematic/enum.ConfigError.html) with a failure
message.
