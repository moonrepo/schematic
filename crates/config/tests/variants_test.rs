use schematic::*;
use serial_test::serial;
use std::collections::HashMap;
use std::env;

fn test_list<T>(_: &[String], _: &T, context: &Context) -> Result<(), ValidateError> {
    if context.fail {
        return Err(ValidateError::new("invalid"));
    }

    Ok(())
}

fn test_map<T>(_: &HashMap<String, String>, _: &T, context: &Context) -> Result<(), ValidateError> {
    if context.fail {
        return Err(ValidateError::new("invalid"));
    }

    Ok(())
}

fn test_cfg<T>(_: &PartialProjectsConfig, _: &T, context: &Context) -> Result<(), ValidateError> {
    if context.fail {
        return Err(ValidateError::new("invalid"));
    }

    Ok(())
}

fn assert_validation_error(error: ConfigError, count: usize) {
    if let ConfigError::Validator { error: inner, .. } = error {
        assert_eq!(inner.len(), count);
    } else {
        panic!("expected validation error");
    }
}

#[derive(Default)]
pub struct Context {
    fail: bool,
}

#[derive(Debug, Config, Eq, PartialEq)]
#[config(context = Context)]
pub struct ProjectsConfig {
    #[setting(validate = test_list)]
    list: Vec<String>,
    #[setting(validate = test_map)]
    map: HashMap<String, String>,
}

#[derive(Debug, Config, Eq, PartialEq)]
#[config(context = Context, serde(untagged))]
pub enum Projects {
    #[setting(nested, validate = test_cfg)]
    Config(ProjectsConfig),
    #[setting(merge = merge::prepend_vec, validate = test_list)]
    List(Vec<String>),
    #[setting(default, merge = merge::merge_hashmap, validate = test_map)]
    Map(HashMap<String, String>),
}

#[derive(Debug, Config, Eq, PartialEq)]
#[config(context = Context)]
pub struct StandardSettings {
    #[setting(nested)]
    projects: Projects,
}

#[test]
#[serial]
fn returns_defaults() {
    let config = ConfigLoader::<StandardSettings>::new()
        .load()
        .unwrap()
        .config;

    assert_eq!(config.projects, Projects::Map(HashMap::new()));
}

mod loading {
    use super::*;

    #[test]
    fn loads_list() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": ["foo", "bar", "baz"]
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::List(vec!["foo".into(), "bar".into(), "baz".into()])
        );
    }

    #[test]
    fn loads_map() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"foo": "bar"
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::Map(HashMap::from_iter([("foo".into(), "bar".into())]))
        );
    }

    #[test]
    fn loads_config() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"list": ["baz"],
		"map": {
			"foo": "bar"
		}
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::Config(ProjectsConfig {
                list: vec!["baz".into()],
                map: HashMap::from_iter([("foo".into(), "bar".into())])
            })
        );
    }
}

mod merging {
    use super::*;

    #[test]
    fn merges_list() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": ["foo", "bar"]
}"#,
                Format::Json,
            )
            .unwrap()
            .code(
                r#"
{
	"projects": ["baz", "qux"]
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::List(vec!["baz".into(), "qux".into(), "foo".into(), "bar".into()])
        );
    }

    #[test]
    fn merges_map() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"foo": "bar"
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .code(
                r#"
{
	"projects": {
		"baz": "qux"
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::Map(HashMap::from_iter([
                ("foo".into(), "bar".into()),
                ("baz".into(), "qux".into())
            ]))
        );
    }

    #[test]
    fn replaces_config() {
        let config = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"list": ["qux", "bar"],
		"map": {
			"baz": "foo"
		}
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load()
            .unwrap()
            .config;

        assert_eq!(
            config.projects,
            Projects::Config(ProjectsConfig {
                list: vec!["qux".into(), "bar".into()],
                map: HashMap::from_iter([("baz".into(), "foo".into())])
            })
        );
    }
}

mod validating {
    use super::*;

    #[test]
    fn validates_list() {
        let error = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": ["foo", "bar", "baz"]
}"#,
                Format::Json,
            )
            .unwrap()
            .load_with_context(&Context { fail: true })
            .err()
            .unwrap();

        assert_validation_error(error, 1);
    }

    #[test]
    fn validates_map() {
        let error = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"foo": "bar"
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load_with_context(&Context { fail: true })
            .err()
            .unwrap();

        assert_validation_error(error, 1);
    }

    #[test]
    fn validates_config() {
        let error = ConfigLoader::<StandardSettings>::new()
            .code(
                r#"
{
	"projects": {
		"list": ["baz"],
		"map": {
			"foo": "bar"
		}
	}
}"#,
                Format::Json,
            )
            .unwrap()
            .load_with_context(&Context { fail: true })
            .err()
            .unwrap();

        // Also validates nested fields
        assert_validation_error(error, 3);
    }
}

#[cfg(feature = "json_schema")]
#[test]
fn generates_json_schema() {
    use starbase_sandbox::{assert_snapshot, create_empty_sandbox};

    let sandbox = create_empty_sandbox();
    let file = sandbox.path().join("schema.json");

    let mut generator = schema::SchemaGenerator::default();
    generator.add::<StandardSettings>();
    generator
        .generate(&file, schema::json_schema::JsonSchemaRenderer::default())
        .unwrap();

    assert!(file.exists());
    assert_snapshot!(std::fs::read_to_string(file).unwrap());
}

#[cfg(feature = "typescript")]
#[test]
fn generates_typescript() {
    use starbase_sandbox::{assert_snapshot, create_empty_sandbox};

    let sandbox = create_empty_sandbox();
    let file = sandbox.path().join("config.ts");

    let mut generator = schema::SchemaGenerator::default();
    generator.add::<StandardSettings>();
    generator
        .generate(&file, schema::typescript::TypeScriptRenderer::default())
        .unwrap();

    assert!(file.exists());
    assert_snapshot!(std::fs::read_to_string(file).unwrap());
}
