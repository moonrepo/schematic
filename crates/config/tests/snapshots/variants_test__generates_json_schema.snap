---
source: crates/config/tests/variants_test.rs
expression: "std::fs::read_to_string(file).unwrap()"
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "StandardSettings",
  "type": "object",
  "required": [
    "projects"
  ],
  "properties": {
    "projects": {
      "allOf": [
        {
          "$ref": "#/definitions/Projects"
        }
      ]
    }
  },
  "additionalProperties": false,
  "definitions": {
    "Projects": {
      "title": "Projects",
      "anyOf": [
        {
          "$ref": "#/definitions/ProjectsConfig"
        },
        {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "propertyNames": {
            "type": "string"
          }
        }
      ]
    },
    "ProjectsConfig": {
      "title": "ProjectsConfig",
      "type": "object",
      "required": [
        "list",
        "map"
      ],
      "properties": {
        "list": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
        "map": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "propertyNames": {
            "type": "string"
          }
        }
      },
      "additionalProperties": false
    }
  }
}

